sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java

```diff
@@ -553,14 +554,20 @@ private static ReadableInstant truncateToMillis(Object value) {
     return new Instant((long) value / 1000);
   }
 
-  private static Object convertAvroArray(Field beamField, Object value) {
+  private static Object convertAvroArray(
+      Field beamField, Object value, BigQueryUtils.ConversionOptions options) {
     // Check whether the type of array element is equal.
     List<Object> values = (List<Object>) value;
     List<Object> ret = new ArrayList();
     for (Object v : values) {
-      ret.add(
-          convertAvroPrimitiveTypes(
-              beamField.getType().getCollectionElementType().getTypeName(), v));
+      FieldType arrayElementType = beamField.getType().getCollectionElementType();
```

Comment:
This, and a little more below, can move out of the loop. Actually you can move it out of the method entirely, I think. You don't need to pass the whole `Field` but it should only depend on the type, right?

Reply:
+1

values actually share the same type because it is an ARRAY. At least collection element type can be gotten out of loop